vulture_os/toolkit/api_parser/beyondtrust_pra/beyondtrust_pra.py

```diff
@@ -0,0 +1,278 @@
+            formated_logs.append(sessions_logs)
+        last_datetime = [x['@timestamp'] for x in formated_logs]
+        # ISO8601 timestamps are sortable as strings
+        last_datetime = max(last_datetime, default=None)
+        return formated_logs, last_datetime
+
+    @staticmethod
+    def format_vault_account_activity_list_logs(logs):
+        formated_logs = logs['vault_account_activity_list'].get('vault_account_activity', [])
+        if isinstance(formated_logs, dict):
+            formated_logs = [formated_logs]
+        last_datetime = [x['@timestamp'] for x in formated_logs]
+        # ISO8601 timestamps are sortable as strings
+        last_datetime = max(last_datetime, default=None)
+        return formated_logs, last_datetime
+
+    def get_logs(self, resource: str, requested_type: str, since: object):
+        url = f"{self.beyondtrust_pra_host}/api/{resource}"
+
+        parameters = {
+            'start_time': int(since.timestamp()),
+            'duration': self.DURATION,
+            'generate_report': requested_type,
+        }
+
+        res = self._execute_query(url, query=parameters)
+
+        logs = xmltodict.parse(res.text)
+
+        if error := logs.get('error'):
+            logger.info(f"[{__parser__}]:get_logs: Error while getting '{requested_type}' logs : {error.get('#text')}", extra={'frontend': str(self.frontend)})
+            return [], None
+
+        # Return only list of events
+        if resource == 'reporting' and requested_type == "Team":
+            return self.format_team_logs(logs)
+        elif resource == 'reporting' and requested_type == "AccessSession":
+            return self.format_access_session_logs(logs)
+        elif resource == 'reporting' and requested_type == "VaultAccountActivity":
+            return self.format_vault_account_activity_list_logs(logs)
+        else:
+            logger.info(f"[{__parser__}]:get_logs: Error while getting '{requested_type}' logs : This requested type is not allowed.", extra={'frontend': str(self.frontend)})
+            return [], None
+
+    def remove_hash_from_keys(self, d):
+        if isinstance(d, dict):
+            return {key.replace('#', ''): self.remove_hash_from_keys(value) for key, value in d.items()}
+        elif isinstance(d, list):
+            return [self.remove_hash_from_keys(item) for item in d]
+        else:
+            return d
+
+    def format_log(self, log, resource, requested_type):
+        log["resource"] = resource
+        log["requested_type"] = requested_type
+        cleaned_log = self.remove_hash_from_keys(log)
+        return json.dumps(cleaned_log)
+
+    def execute(self):
+
+        # Get reporting logs
+        reporting_types = ["Team", "AccessSession", "VaultAccountActivity"]
+
+        for report_type in reporting_types:
+            since = self.last_collected_timestamps.get(f"beyondtrust_pra_{report_type}") or (timezone.now() - timedelta(days=30))
```

Comment:
please add a 'to' parameter to bound the request
warning: it seems (see [documentation](https://www.beyondtrust.com/docs/privileged-remote-access/how-to/integrations/api/reporting/access-session.htm)) that query filtering **INCLUDES** both since and to in the range, so 'to' value will have to be slightly incremented once the process is finished (or 'since' at the beggining of processing, which may be better)

Reply:
the since parameter is provided by default to work with the "duration" argument.